src/app/clusters/time-synchronization-server/time-synchronization-server.h

```diff
@@ -0,0 +1,170 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file API declarations for time sync cluster.
+ */
+
+#pragma once
+
+#include "TimeSyncDataProvider.h"
+
+#include <app/server/Server.h>
+#include <app/util/af-types.h>
+#include <app/util/config.h>
+#include <lib/core/TLV.h>
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TimeSynchronization {
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeSynchronization;
+using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+using chip::TimeSyncDataProvider;
+using chip::Protocols::InteractionModel::Status;
+
+class TimeSynchronizationServer
+{
+    struct timeZoneName
+    {
+        char name[64];
+    };
+
+public:
+    void Init();
+
+    static TimeSynchronizationServer & Instance(void);
+
+    TimeSyncDataProvider GetDataProvider(void) { return mTimeSyncDataProvider; }
+
+    CHIP_ERROR SetTrustedTimeSource(DataModel::Nullable<TimeSynchronization::Structs::TrustedTimeSourceStruct::Type> tts)
+    {
+        if (!tts.IsNull())
+        {
+            mTrustedTimeSource.SetNonNull(tts.Value());
+            mTimeSyncDataProvider.StoreTrustedTimeSource(tts.Value());
+        }
+        else
+        {
+            mTrustedTimeSource.SetNull();
+            mTimeSyncDataProvider.ClearTrustedTimeSource();
+        }
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR SetDefaultNtp(DataModel::Nullable<chip::MutableByteSpan> dntp)
+    {
+        if (!dntp.IsNull())
+        {
+            memcpy(mDefaultNtpBuf, dntp.Value().data(), dntp.Value().size());
+            mDefaultNtp.SetNonNull(MutableByteSpan(mDefaultNtpBuf, dntp.Value().size()));
+            return mTimeSyncDataProvider.StoreDefaultNtp(mDefaultNtp.Value());
+        }
+        else
+        {
+            mDefaultNtp.SetNull();
+            return mTimeSyncDataProvider.ClearDefaultNtp();
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR SetTimeZone(DataModel::DecodableList<TimeSynchronization::Structs::TimeZoneStruct::Type> tz)
+    {
+        auto mTzL = mTimeZoneList.begin();
+        auto tzL  = tz.begin();
+        size_t i  = 0;
+
+        while (tzL.Next())
+        {
+            mTzL[i].offset  = tzL.GetValue().offset;
+            mTzL[i].validAt = tzL.GetValue().validAt;
+            if (tzL.GetValue().name.HasValue())
+            {
+                const char * buf = tzL.GetValue().name.Value().data();
+                size_t len       = tzL.GetValue().name.Value().size();
+                Platform::CopyString(mNames[i].name, chip::CharSpan(buf, len));
```

Comment:
This is going to fail, no?  The buffers we are copying into can hold at most 64 chars, but we are trying to put up to 64 chars _and a null terminator_ into them.  We should have a test that checks that you can in fact put a 64-char timezone name in.

Why are we bothering with a null terminator?  What ends up using it?

Reply:
Are you suggesting to use direct memory copy? 
Where to write test for such things?

Reply:
> Are you suggesting to use direct memory copy?

Yes.

>  Where to write test for such things?

This should be doable in a YAML test, for example.

Reply:
These internal things can be done using the NL test suite. There's a bunch of examples in the test/ directories. Ex. lib/support/tests/TestSpan.cpp does a bunch of memory compare tests, as does TestSafeString.cpp. I'd actually start with those as it's more of a unit testing framework. The yaml tests tend to be more for server-client interactions, which is why we use them for cert.

Reply:
I have already updated this implementation. Not sure if I can do more here in relation to testing.

Reply:
why would this be hard with NL tests? It's a public API, should be fairly easy to test. Are you concerned about the integration with the attribute store?

Reply:
Yes, the hard part will be making sure all the scaffolding is in place to make it actually work.  But I'd be happy to be proved wrong!

Reply:
For example, to call this public API you have to encode some data to TLV, then set up a DecodableList pointing at that TLV, etc.  It's not _that_ hard if you know what all the pieces are, I guess.  And it might be enough to just do those pieces.

Reply:
We have a 65-char test in the cert test and python impl. Added an issue to add a check for 64-bit exactly. Test plan issue https://github.com/CHIP-Specifications/chip-test-plans/issues/2976. I will follow up with that change and the change to the python scripts. Resolving this conversation here, if further discussion is needed please comment on the test plan issue.